Quotient/connection.cpp

```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+    if (result->has_error()) {
+        return false;
+    }
+    return result->value();
 }
 
-QString Connection::curveKeyForUserDevice(
-    const QString& userId, const QString& device) const
+void Connection::Private::runShareKey(unsigned int requestId, Room* room, QPromise<void>&& promise)
+{
+    auto historyVisibilityEvent = room->currentState().get("m.room.history_visibility"_L1);
+    int visibility = 0;
+    if (historyVisibilityEvent) {
+        auto visibilityString = historyVisibilityEvent->contentJson()["history_visibility"_L1].toString();
+        if (visibilityString == u"joined"_s) {
+            visibility = 1;
+        } else if (visibilityString == u"shared"_s) {
+            visibility = 2;
+        } else if (visibilityString == u"world_readable"_s) {
+            visibility = 3;
+        }
+
+    }
+
+    rust::Vec<rust::String> ids;
+    for (const auto& id : room->joinedMemberIds()) {
+        ids.push_back(stringToRust(id));
+    }
+
+    if (visibility == 0 || visibility == 3) {
+        for (const auto& event : room->currentState().eventsOfType(u"m.room.member"_s)) {
+            if (auto memberEvent = eventCast<const RoomMemberEvent>(event)) {
+                if (memberEvent->isInvite()) {
+                    ids.push_back(stringToRust(memberEvent->stateKey()));
+                }
+            }
+        }
+    }
+
+    auto missingResult = (*cryptoMachine)->get_missing_sessions(ids);
+    if (missingResult->has_error() || !missingResult->has_request()) {
+        Q_EMIT q->shareRoomKeyDone(requestId);
+        if (lastScheduledRequest == requestId) {
+            lastScheduledRequest = 0;
+        }
+        return;
+    }
+    auto missing = missingResult->request();
+    auto id = missing->id();
+    q->callApi<ClaimKeysJob>(fromJson<QHash<UserId, QHash<QString, QString>>>(jsonFromRust(missing->one_time_keys()))).then([this, visibility, id, room, ids, p = std::move(promise), requestId](const auto& job) mutable {
+        (*cryptoMachine)->mark_keys_claim_as_sent(bytesToRust(job->rawData()), id);
+
+        Q_EMIT q->shareRoomKeyDone(requestId);
+        if (lastScheduledRequest == requestId) {
+            lastScheduledRequest = 0;
+        }
+        auto requestsResult = (*cryptoMachine)->share_room_key(stringToRust(room->id()), ids, false, visibility);
+        if (requestsResult->has_error()) {
+            return;
+        }
+        auto requests = requestsResult->value();
+        p.finish();
+        for (const auto& request : requests) {
+            auto id = request.txn_id();
+
+            q->callApi<SendToDeviceJob>(stringFromRust(request.event_type()), stringFromRust(id), fromJson<QHash<UserId, QHash<QString, QJsonObject>>>(jsonFromRust(request.messages()))).then([this, id](const auto& job) {
+                (*cryptoMachine)->mark_to_device_as_sent(bytesToRust(job->rawData()), id);
+            }, [=](const auto& job){});
```

Comment:
This is the default value for the second parameter of `JobHandle::then`, you shouldn't need to spell it out, unless `=` does something non-trivial (doubt it).
```suggestion
            });
```